homeassistant/components/travel_time/__init__.py

```diff
@@ -0,0 +1,202 @@
+"""The travel_time integration."""
+from datetime import timedelta
+import logging
+from typing import Dict, Optional, Union
+
+import voluptuous as vol
+
+from homeassistant.const import ATTR_ATTRIBUTION
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_component import EntityComponent
+
+from .const import (
+    ATTR_DESTINATION,
+    ATTR_DESTINATION_ADDRESS,
+    ATTR_DISTANCE,
+    ATTR_DURATION,
+    ATTR_DURATION_IN_TRAFFIC,
+    ATTR_ORIGIN,
+    ATTR_ORIGIN_ADDRESS,
+    ATTR_ROUTE,
+    ATTR_TRAVEL_MODE,
+    CONF_DESTINATION_LATITUDE,
+    CONF_DESTINATION_LONGITUDE,
+    CONF_DESTINATION_NAME,
+    CONF_ORIGIN_LATITUDE,
+    CONF_ORIGIN_LONGITUDE,
+    CONF_ORIGIN_NAME,
+    DOMAIN,
+    UNIT_OF_MEASUREMENT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+SCAN_INTERVAL = timedelta(minutes=5)
+
+TRAVEL_TIME_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Inclusive(CONF_DESTINATION_LATITUDE, "destination_coordinates"): vol.Any(
+            cv.latitude, cv.template
+        ),
+        vol.Inclusive(CONF_DESTINATION_LONGITUDE, "destination_coordinates"): vol.Any(
+            cv.longitude, cv.template
+        ),
+        vol.Inclusive(CONF_ORIGIN_LATITUDE, "origin_coordinates"): vol.Any(
+            cv.latitude, cv.template
+        ),
+        vol.Inclusive(CONF_ORIGIN_LONGITUDE, "origin_coordinates"): vol.Any(
+            cv.longitude, cv.template
+        ),
+        vol.Exclusive(CONF_DESTINATION_LATITUDE, "destination"): vol.Any(
+            cv.latitude, cv.template
+        ),
+        vol.Exclusive(CONF_DESTINATION_NAME, "destination"): vol.Any(
+            cv.string, cv.template
+        ),
+        vol.Exclusive(CONF_ORIGIN_LATITUDE, "origin"): vol.Any(
+            cv.latitude, cv.template
+        ),
+        vol.Exclusive(CONF_ORIGIN_NAME, "origin"): vol.Any(cv.string, cv.template),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORM_SCHEMA = vol.All(
+    cv.has_at_least_one_key(CONF_DESTINATION_LATITUDE, CONF_DESTINATION_NAME),
+    cv.has_at_least_one_key(CONF_ORIGIN_LATITUDE, CONF_ORIGIN_NAME),
+    TRAVEL_TIME_SCHEMA,
+)
+
+
+async def async_setup(hass, config):
+    """Track states and offer events for sensors."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+
+    await component.async_setup(config)
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class TravelTimeEntity(Entity):
```

Comment:
I've been thinking about this more and I think that we should have platforms implement entities. Shouldn't it just be services that take in a start, end and options and return a measurement which then is represented by an entity.

Reply:
The idea behind this was the other way round. Define a clear set of properties a TravelTime Entity should have so they can be used in Lovelace cards, automations, appdaemon etc.
Currently all of the above look different for Google, Waze, Here....
For example: 

Google has an attribute `destination_addresses` which is a list.
Waze has `destination_address` HERE has `destination_name`.

Additional thought was that the configuration options should be named the same.

Reply:
That would still work ? The `travel_time` integration defines thte entity and the exposed data, so it's the same. It just sources that data from a service object provided by integrations instead of having integrations provide the entity objects directly.

Reply:
Is there an existing platform/integration which does that already so I could have a look?

Reply:
Notification, but I don't like that implementation :( 

I guess we can still use entities but just have each entity implement an abstract method that returns all info at once ? 

I don't know. Maybe the existing approach is ok. 🤷‍♂ It just feels a little weird.